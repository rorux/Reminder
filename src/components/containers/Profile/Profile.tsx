import React, { useCallback, useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { CircularProgress } from '@mui/material';
import Typography from '@mui/material/Typography';
import TextField from '@mui/material/TextField';
import Button from '@mui/material/Button';
import { getProfileFromFirebase, changeProfileWithFirebase } from '@store/profile/actions';
import { profileSelector } from '@store/profile/selectors';

const Profile = () => {
  const [name, setName] = useState('');
  const { userName, loading } = useSelector(profileSelector);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getProfileFromFirebase());
  }, [dispatch]);

  const changeName = (event: React.ChangeEvent<HTMLInputElement>) => {
    setName(event?.target.value as string);
  };

  const changeNameWithFirebase = useCallback(() => {
    dispatch(changeProfileWithFirebase({ userName: name }));
  }, [dispatch, name]);

  return (
    <>
      {loading ? (
        <CircularProgress />
      ) : (
        <div data-test="profile">
          <Typography variant="body1">
            Ваше имя: <strong>{userName}</strong>
          </Typography>
          <TextField
            id="standard-basic"
            label="Введите другое имя.."
            variant="standard"
            sx={{ marginTop: 2, marginRight: 2 }}
            value={name}
            onChange={changeName}
          />
          <Button
            disabled={!name}
            variant="contained"
            color="success"
            sx={{ marginTop: 3 }}
            onClick={changeNameWithFirebase}
          >
            Изменить имя
          </Button>
        </div>
      )}
    </>
  );
};

export default Profile;
