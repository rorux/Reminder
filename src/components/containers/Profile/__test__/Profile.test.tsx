import React from 'react';
import * as Redux from 'react-redux';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import 'jsdom-global/register';
import Enzyme, { shallow } from 'enzyme';
import { CircularProgress } from '@mui/material';
import Button from '@mui/material/Button';
import TextField from '@mui/material/TextField';
import Profile from '../Profile';

Enzyme.configure({ adapter: new Adapter() });

const createSelector = (loading: boolean) => ({
  userName: null,
  loading,
  error: false,
});

describe('Profile component', () => {
  let useEffect: any;
  let useDispatch: any;
  let useSelector: any;
  let setState: any;
  let useStateSpy;
  let mockDispatch: jest.Mock;

  const mockUseEffect = () => {
    useEffect.mockImplementationOnce((f: any) => f());
  };

  beforeEach(() => {
    useEffect = jest.spyOn(React, 'useEffect');
    useDispatch = jest.spyOn(Redux, 'useDispatch');
    useSelector = jest.spyOn(Redux, 'useSelector');
    mockDispatch = jest.fn();
    useDispatch.mockReturnValue(mockDispatch);
    setState = jest.fn();
    useStateSpy = jest.spyOn(React, 'useState');
    // @ts-ignore
    useStateSpy.mockImplementation((init) => [init, setState]);
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('Snapshot of Dashboard component', () => {
    useSelector.mockReturnValue(createSelector(false));
    const component = shallow(<Profile />);
    expect(component).toMatchSnapshot();
  });

  it('Should work "useEffect"', () => {
    useSelector.mockReturnValue(createSelector(false));
    mockUseEffect();
    shallow(<Profile />);
    expect(mockDispatch).toHaveBeenCalled();
  });

  it('Data entry should change state', () => {
    useSelector.mockReturnValue(createSelector(false));
    const component = shallow(<Profile />);
    // @ts-ignore
    component.find(TextField).props().onChange();
    expect(setState).toHaveBeenCalled();
  });

  it('Click button should dispatch action', () => {
    useSelector.mockReturnValue(createSelector(false));
    const component = shallow(<Profile />);
    // @ts-ignore
    component.find(Button).props().onClick();
    expect(mockDispatch).toHaveBeenCalled();
  });

  it('Should render CircularProgress when loading', () => {
    useSelector.mockReturnValue(createSelector(true));
    const component = shallow(<Profile />);
    expect(component.find(CircularProgress)).toHaveLength(1);
  });
});
