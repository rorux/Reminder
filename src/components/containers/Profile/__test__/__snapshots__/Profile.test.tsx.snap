// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile component Snapshot of Dashboard component 1`] = `
<Fragment>
  <div
    data-test="profile"
  >
    <ForwardRef(Typography)
      variant="body1"
    >
      Ваше имя: 
      <strong />
    </ForwardRef(Typography)>
    <ForwardRef(TextField)
      id="standard-basic"
      label="Введите другое имя.."
      onChange={[Function]}
      sx={
        Object {
          "marginRight": 2,
          "marginTop": 2,
        }
      }
      value=""
      variant="standard"
    />
    <ForwardRef(Button)
      color="success"
      disabled={true}
      onClick={[Function]}
      sx={
        Object {
          "marginTop": 3,
        }
      }
      variant="contained"
    >
      Изменить имя
    </ForwardRef(Button)>
  </div>
</Fragment>
`;
